// ng2-image-lazy-load/src/client/app/providers/image-lazy-load.provider.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var http_1 = require('@angular/http');
var web_worker_provider_1 = require('./web-worker.provider');
var ImageLazyLoaderService = (function () {
    function ImageLazyLoaderService(http, worker) {
        this.http = http;
        this.worker = worker;
        this.imageCache = {};
    }
    /*
    * Loads the url via `WebWorker` where supported and gracefully degrades to using `Http` if needed.
    * @param url  url of image to load.
    * @param headers  **(optional)** any custom headers (as an `Object`) that may be required to load the image.
    * @returns return `Promise`
    */
    ImageLazyLoaderService.prototype.load = function (url, headers) {
        if (this.imageCache[url]) {
            // image has been previously loaded, rely on browser cache
            return Promise.resolve(true);
        }
        else if (web_worker_provider_1.WebWorkerService.supported) {
            return this.loadViaWorker(url, headers);
        }
        else {
            return this.loadViaHttp(url, headers);
        }
    };
    /*
    * Loads the url via `WebWorker` directly.
    * @param url  url of image to load.
    * @param headers  **(optional)** any custom headers (as an `Object`) that may be required to load the image.
    * @returns return `Promise`
    */
    ImageLazyLoaderService.prototype.loadViaWorker = function (url, headers) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var id, completeHandler, msgFn, errorFn;
            completeHandler = function (success, err) {
                _this.worker.terminate(id);
                if (success) {
                    _this.imageCache[url] = true;
                    resolve(true);
                }
                else {
                    reject(err);
                }
            };
            msgFn = function (e) {
                if (e && e.data !== 'ERROR') {
                    completeHandler(true);
                }
                else {
                    completeHandler(false, e);
                }
            };
            errorFn = function (e) {
                completeHandler(false, e);
            };
            var config = {
                method: 'GET',
                url: url
            };
            // optionally set headers
            if (headers) {
                config.headers = headers;
            }
            id = _this.worker.load(config, msgFn, errorFn);
        });
    };
    /*
    * Loads the url via `Http` directly.
    * @param url  url of image to load.
    * @param headers  **(optional)** any custom headers (as an `Object`) that may be required to load the image.
    * @returns return `Promise`
    */
    ImageLazyLoaderService.prototype.loadViaHttp = function (url, headers) {
        var _this = this;
        var ro = null;
        if (headers) {
            ro = new http_1.RequestOptions({
                headers: new http_1.Headers(headers)
            });
        }
        return new Promise(function (resolve, reject) {
            _this.http.get(url, ro)
                .subscribe(function (res) {
                _this.imageCache[url] = true;
                resolve(true);
            });
        });
    };
    ImageLazyLoaderService = __decorate([
        core_1.Injectable(), 
        __metadata('design:paramtypes', [http_1.Http, web_worker_provider_1.WebWorkerService])
    ], ImageLazyLoaderService);
    return ImageLazyLoaderService;
}());
exports.ImageLazyLoaderService = ImageLazyLoaderService;
