// ng2-image-lazy-load/src/client/app/providers/web-worker.provider.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var WebWorkerService = (function () {
    function WebWorkerService() {
        this.activeWorkers = [];
    }
    WebWorkerService.prototype.load = function (config, msgFn, errorFn) {
        if (typeof (config) !== 'object') {
            throw ("config must be an Object with method and url defined.");
        }
        else if (!config.url) {
            throw ("config.url must be defined.");
        }
        var id = Math.floor(Math.random() * 1000000000000);
        var w = new Worker(WebWorkerService.workerUrl);
        if (msgFn) {
            w.addEventListener('message', msgFn, false);
        }
        if (errorFn) {
            w.addEventListener('error', errorFn, false);
        }
        this.activeWorkers.push({
            id: id,
            worker: w,
            msgFn: msgFn,
            errorFn: errorFn
        });
        w.postMessage(config);
        return id;
    };
    WebWorkerService.prototype.terminate = function (id) {
        var index = this.activeWorkers.findIndex(function (item) { return item.id === id; });
        if (index > -1) {
            var activeWorker = this.activeWorkers[index];
            if (activeWorker && activeWorker.worker) {
                if (activeWorker.worker.msgFn) {
                    activeWorker.worker.removeEventListener('message', activeWorker.worker.msgFn);
                }
                if (activeWorker.worker.errorFn) {
                    activeWorker.worker.removeEventListener('error', activeWorker.worker.errorFn);
                }
                activeWorker.worker.terminate();
            }
            this.activeWorkers.splice(index, 1);
        }
    };
    WebWorkerService.supported = typeof (Worker) !== 'undefined';
    WebWorkerService.workerUrl = 'assets/js/xhrWorker.js';
    WebWorkerService = __decorate([
        core_1.Injectable(), 
        __metadata('design:paramtypes', [])
    ], WebWorkerService);
    return WebWorkerService;
}());
exports.WebWorkerService = WebWorkerService;
