// @ngrx/notify/ng2.js
"use strict";
var core_1 = require('@angular/core');
var notify_1 = require('./notify');
var notification_permission_1 = require('./notification-permission');
exports.NOTIFY_GLOBAL_OPTIONS = new core_1.OpaqueToken('[@ngrx/notify] Global Options');
exports.NOTIFICATION_STATIC = new core_1.OpaqueToken('[@ngrx/notify] Notification Static Constructor');
var NOTIFICATION_STATIC_PROVIDER = {
    provide: exports.NOTIFICATION_STATIC,
    useValue: Notification
};
var NOTIFICATION_PERMISSION_PROVIDER = {
    provide: notification_permission_1.NotificationPermission,
    deps: [exports.NOTIFICATION_STATIC],
    useFactory: function (Notification) {
        return new notification_permission_1.NotificationPermission(Notification);
    }
};
var NOTIFY_PROVIDER = {
    provide: notify_1.Notify,
    deps: [exports.NOTIFICATION_STATIC, exports.NOTIFY_GLOBAL_OPTIONS, notification_permission_1.NotificationPermission],
    useFactory: function (Notification, options, permission$) {
        return new notify_1.Notify(Notification, options, permission$);
    }
};
var DEFAULT_GLOBAL_OPTIONS_PROVIDER = {
    provide: exports.NOTIFY_GLOBAL_OPTIONS,
    multi: true,
    useValue: {}
};
exports.NOTIFY_PROVIDERS = [
    NOTIFICATION_STATIC_PROVIDER,
    NOTIFICATION_PERMISSION_PROVIDER,
    NOTIFY_PROVIDER,
    DEFAULT_GLOBAL_OPTIONS_PROVIDER
];
